File: swift/problem501.swift
public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init(_ val: Int) {
        self.val = val
        self.left = nil
        self.right = nil
    }
}

//swift implementation of cpp version
class Solution {
    func findMode(_ root: TreeNode?) -> [Int] {
        if (root == nil) {
            return []
        }
        
        var mode = Int.min; var modeCount = 0; var MaxCount = 0
        var ans = [Int]()
        
        //using closure
        let findMaxCount = { (_ val: Int) -> () in
            if (val == mode) {
                modeCount += 1
            } else {
                MaxCount = max(MaxCount,modeCount)
                mode = val
                modeCount = 1
            }
        }
        inOrder(root,findMaxCount)
        MaxCount = max(MaxCount,modeCount)
        
        mode = Int.min; modeCount = 0
        
        let findAns = { (_ val: Int) -> () in
            if (val == mode) {
                modeCount += 1
            } else {
                if (modeCount == MaxCount) {
                    ans.append(mode)
                }
                mode = val
                modeCount = 1
            }
        }
        inOrder(root,findAns)
        if (modeCount == MaxCount) {
            ans.append(mode)
        }
        
        return ans
    }
    
    func inOrder(_ root: TreeNode?, _ function: (_ val: Int) -> ()) {
        var p = root; var stack = [TreeNode]()
        
        while (p != nil || !stack.isEmpty) {
            while (p != nil) {
                stack.append(p!)
                p = p!.left
            }
            p = stack.removeLast()
            function(p!.val)
            p = p!.right
        }
    }
}

File: cpp/problem501.cpp
#include <mystd.h>
#include "BinaryTree.h"
using namespace MYSTD;
using namespace std;

typedef MYDS::BinaryTree<int> Tree;
typedef typename Tree::node_type TreeNode;
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */


/**
 * Solution by in-order DFS
 * 
 * To minimize the memory cost, we find the modes by two runs of DFS.
 * The first run gives number of occurance of mode elements.
 * The second run picks up the modes.
 *
 * Space: O(tree.height); Time: Theta(tree.size).
 */
class Solution {
public:
    vector<int> findMode(TreeNode* root) {
        if( !root ) return {};

        // first run of in-order, find number of occurance of mode elements
        int mode = INT_MIN, cnt_mode = 0, max_cnt_mode = 0;
        auto find_cnt_mode = [&mode, &cnt_mode, &max_cnt_mode](TreeNode *p){
            if( p->val == mode ) ++cnt_mode;
            else{
                max_cnt_mode = max(max_cnt_mode, cnt_mode);
                mode = p->val;
                cnt_mode = 1;
            }
        };
        inOrder( root, find_cnt_mode );
        max_cnt_mode = max(max_cnt_mode, cnt_mode);
        
        // second run of in-order, pick up modes
        vector<int> res;
        mode = INT_MIN, cnt_mode = 0;
        auto give_cnt_mode = [&mode, &cnt_mode, &max_cnt_mode, &res](TreeNode *p){
            if( p->val == mode ) ++cnt_mode;
            else{
                if( cnt_mode == max_cnt_mode ) res.push_back( mode );
                mode = p->val;
                cnt_mode = 1;
            }
        };
        inOrder( root, give_cnt_mode );
        if( cnt_mode == max_cnt_mode ) res.push_back(mode);

        return res;
    }
    /**
     * in-order visiting by any function f
     */ 
    void inOrder(TreeNode* root, function<void(TreeNode *)> f) {
        stack<TreeNode *> stk;
        TreeNode *p = root;
        while( p || !stk.empty() ){
            while(p){
                stk.emplace(p);
                p = p->left;
            }
            p = stk.top(); stk.pop();
            f(p);
            p = p->right;
        }
    }
};


int main(int argc, char const *argv[]){
    // test case in Leetcode Problem
    vector<int> v = {1,-1,2,2};
    Tree tree(v, -1);
    cout << tree;

    Solution sol;
    auto res = sol.findMode( tree.root() );
    MYFILE::print_range( res ) << endl;

    return 0;
}
